### PR TITLE
Don't log empty m2m changes (fix #2576)

### LINKED ISSUE #2576: Don't log empty m2m changes
Currently, also empty m2m changes are logged, e. g. when calling `course.programs.add(*[])`. This should bef

Example:

```
INSERT INTO "evaluation_logentry" ("content_type_id", "content_object_id", "attached_to_object_type_id", "attached_to_object_id", "datetime", "user_id", "action_type", "request_id", "data") VALUES (17, 115, 17, 115, '2025-12-01 20:44:15.381554'::timestamp, NULL, 'change', '', '{"programs": {"add": []}}'::jsonb) RETURNING "evaluation_logentry"."id"
```

### PR DESCRIPTION
**Pull request by LukasMansour**
See #2576

### CODE REVIEW

File: evap/evaluation/tests/tools.py [hunk 1]
```diff
@@ -294,6 +296,45 @@ def assert_no_database_modifications(*args, **kwargs):
...
+            if query["sql"].startswith('INSERT INTO "evaluation_logentry"') and has_bad_logging_data(
+                extract_jsonb(query["sql"])
+            ):
+                raise AssertionError("Unexpected modifying query found: " + query["sql"])
+
+
+# Helper method that extracts the json literal from an SQL Values clause.
+# Assumes the JSON appears between the last single quotes.
+# Added to solve issue: #2576
+def extract_jsonb(sql: str):
```
Comment (richardebeling) [inline evap/evaluation/tests/tools.py]: unused? I would hope we don't need to write a broken-in-edge-cases extractor here, [there must be a better way](https://www.reddit.com/r/Python/comments/n54f9s/raymond_hettinger_hitting_the_table_there_must_be/)

File: evap/staff/tests/test_importers.py [hunk 2]
```diff
@@ -341,7 +341,8 @@ def create_existing_course(self) -> tuple[Course, Evaluation]:
         return existing_course, existing_course_evaluation
 
     def test_valid_file_import(self):
-        importer_log = import_enrollments(self.default_excel_content, self.semester, None, None, test_run=True)
+        with assert_no_database_useless_logging():
+            importer_log = import_enrollments(self.default_excel_content, self.semester, None, None, test_run=True)
```
Comment (richardebeling) [inline evap/staff/tests/test_importers.py]: My understanding was that this doesn't do any actual database modifications, but it just "accidentally" added the empty log entries. If this PR fixes the empty log entries, we should be good to use `assert_no_database_modifications` (and hopefully we don't need `assert_no_database_useless_logging` at all, then)

File: evap/evaluation/tests/tools.py [hunk 3]
```diff
@@ -294,6 +296,45 @@ def assert_no_database_modifications(*args, **kwargs):
                 raise AssertionError("Unexpected modifying query found: " + query["sql"])
 
 
+@contextmanager
+def assert_no_database_useless_logging(*args, **kwargs):
```
Comment (richardebeling) [inline evap/evaluation/tests/tools.py]: I think with proper testing setup this shouldn't be necessary. We should:
* have a unit test testing the case from the issue, ` course.programs.add(*[])`. This should pass under `assert_no_database_modifications`
* modify the tests for the importer to use `assert_no_database_modifications` (which we would have liked to do before, but couldn't, due to the logging entries)

File: evap/evaluation/models_logging.py [hunk 4]
```diff
@@ -358,18 +358,25 @@ def _m2m_changed(sender, instance, action, reverse, model, pk_set, **kwargs):  #
...
-                action_type = FieldActionType.M2M_REMOVE
-            case _:
-                return
+    match action:
+        case "pre_remove":
+            action_type = FieldActionType.M2M_REMOVE
+        case "pre_add":
+            action_type = FieldActionType.M2M_ADD
+        case "pre_clear":
+            action_type = FieldActionType.M2M_CLEAR
```
Comment (richardebeling) [inline evap/evaluation/models_logging.py:367]: Can you explain why the changes here are necessary, and how this is now supposed to work for the reverse-clear case? My understanding is that if we do a reverse-clear (i.e., `A.b_instances.clear()`), we want to treat it as N forward removes (i.e., as if `B.a_instances.remove()` was called on each B that was in `A.b_instances`). This, in my head, should still work perfectly fine with a generic "If adding or removing an empty set, bail out early" logic below -- and it should even require it: If `A.b_instances` was already empty, we don't want to trigger logging, but right now, your early-out wouldn't catch this case.

File: evap/evaluation/models_logging.py [hunk 5]
```diff
@@ -358,18 +358,26 @@ def _m2m_changed(sender, instance, action, reverse, model, pk_set, **kwargs):  #
...
+        case _:
+            return
+
+    if action_type == (FieldActionType.M2M_ADD or FieldActionType.M2M_REMOVE):
```
Comment (niklasmohrin) [inline evap/evaluation/models_logging.py]: This does not do what we intend; it needs to be something like
```suggestion
    if action_type in (FieldActionType.M2M_ADD, FieldActionType.M2M_REMOVE):
```

File: evap/evaluation/models_logging.py [hunk 6]
```diff
@@ -358,18 +358,26 @@ def _m2m_changed(sender, instance, action, reverse, model, pk_set, **kwargs):  #
...
+        if not pk_set:
```
Comment (niklasmohrin) [inline evap/evaluation/models_logging.py]: (I also think we can merge the two nested ifs into one)

File: evap/evaluation/tests/test_models_logging.py [hunk 7]
```diff
@@ -205,3 +205,12 @@ def test_m2m_logging_respects_unlogged_fields(self):
             participant.evaluations_participating_in.clear()
             participant.evaluations_participating_in.add(self.evaluation)
             self.assertFalse(any("participants" in entry.data for entry in self.evaluation.related_logentries()))
+
+    def test_no_empty_m2m_logs(self):
+        evaluation = baker.make(Evaluation)
```
Comment (niklasmohrin) [inline evap/evaluation/tests/test_models_logging.py]: what do we need the evaluation for? shouldn't making a course be enough?

**Review by richardebeling — APPROVED** [review]
